batch/batch/server/server.py

```diff
@@ -934,14 +937,32 @@ def is_successful(self):
     return {'batch': batch}
 
 
+@routes.post('/batches/{batch_id}/cancel')
+@aiohttp_jinja2.template('batches.html')
+@authenticated_users_only
+async def ui_cancel_batch(request, userdata):
+    batch_id = int(request.match_info['batch_id'])
+    user = userdata['username']
+
+    token = request.query.get('_csrf')
+    if not await db.tokens.has_token(token):
+        abort(404)
+
+    await _cancel_batch(batch_id, user)
+
+    location = request.app.router['batches'].url_for()
+    raise web.HTTPFound(location=location)
+
+
 @routes.get('/batches', name='batches')
 @aiohttp_jinja2.template('batches.html')
 @authenticated_users_only
 async def ui_batches(request, userdata):
     params = request.query
     user = userdata['username']
     batches = await _get_batches_list(params, user)
-    return {'batch_list': batches}
+    token = await db.tokens.new_token()
```

Comment:
This will accumulate a million tokens, and bring a big vulnerability. 

Imagine I'm a malicious user. I write a bit of javascript that, when triggered by a user in a malicious email, calls /batches 1 million times. I will generate 1 million SQL requests. First, this will represent an easier path to DDOS (crash the database, prevent all future requests). Second, I now have 1 million tokens in the database. If I can guess the key length, My chances of guessing a valid CSRF token have increased by 1 million times.

If I do this for many users, I can effectively always guess a valid key within a 30 minute window.

Reply:
ok. That's why I had you review this. What's the correct thing to do then?